src/core_deku/contract_context.ml

```diff
@@ -0,0 +1,155 @@
+open Core
+
+module Contract_operation = struct
+  type t =
+    | Transfer of {
+        destination : Address.t;
+        ticket : Ticket_id.t;
+        amount : Amount.t;
+      }
+    | Invoke   of {
+        param : bytes;
+        destination : Address.t;
+        ticket : Ticket_id.t;
+      }
+end
+module type CTX = sig
+  val get_table : unit -> Ticket_table.t
+  val get_ops :
+    int list -> (Contract_operation.t list, [`Execution_error]) result
+
+  val get_source : Address.t
+  val sender : unit -> Address.t
+  val source : unit -> Address.t
+  val self : unit -> Address.t
+  val read_ticket : Ticket_handle.t -> Ticket_id.t * Amount.t * Ticket_handle.t
+  val split_ticket :
+    Ticket_handle.t * Amount.t * Amount.t -> Ticket_handle.t * Ticket_handle.t
+
+  val join_tickets : Ticket_handle.t * Ticket_handle.t -> Ticket_handle.t
+  val own_ticket : Ticket_handle.t -> Ticket_handle.t
+  val get_contract_opt : Address.t -> Address.t option
+  val transaction : bytes * (Ticket_handle.t * Amount.t) * Address.t -> int
+end
+module Ctx = struct
+  module S = Helpers.Set.Make (Ticket_handle)
+  let get_or_raise t =
+    Result.map_error
+      ~f:(fun e ->
+        Format.asprintf "%a" Ticket_table.Errors.pp (e :> Ticket_table.Errors.t))
+      t
+    |> Result.ok_or_failwith
+
+  let make ~sender ~table ~self ~source ~tickets ~contracts_table =
+    (module struct
```

Comment:
Why so much mutable state? Why not doing more like Michelson and having stuff as immutable values? It'd be a lot better for testing.

Reply:
Its localised per one contract call, so mutable state is not a problem, because it will be used only once per invocation and state should be kept somewhere should be kept somewhere

Reply:
All right